### PR TITLE
Apt editor

### PR DESCRIPTION
**Pull request by BSG-75**
- Allow Viewing (not really) & Editing of Apt files
Tested (by tested I mean it usually doesn't crash) on: BFME apt files, [RA3 UI Source Pack](https://www.moddb.com/games/cc-red-alert-3/downloads/ra3-mod-ui-source-pack)
#### Things that are (a bit) working:
- Opening Apt files (and resolving their imports, currently not well implemented, but kinda works)
- Creating `Shape` / `Geometry` (geometry handled in a text editor)
- Creating / Editing `PlaceObject` in a `Playable`'s `Frame`
- Exporting Apt files
#### Main things that are still missing:
- Correct geometry transformation
- All Apt Characters expect `Sprite` and `Shape`
- Everything ActionScript related (actually now AptEditor uses a hack to prevent OpenSage from executing ActionScript)
- Creation of new frames, creation of multi-frame interpolation animation (like what Adobe Flash does)
- Edit geometries in a interactive editor instead of text editor
- Edit `ImageMap` / import more custom textures

### CODE REVIEW

**Review by tgjones — APPROVED** [review]
This looks amazing, thank you! I didn’t go through it line by line but in broad brush strokes the code looks good to me.

**Comment by Tarcontar** [comment]
Sadly this causes some issues with the bfme ui.
image

File: src/OpenSage.Game/Gui/Apt/AptRenderingContext.cs [hunk 1]
```diff
@@ -135,7 +135,7 @@ public void RenderText(Text text)
             CalculateTransform(ref transform);
 
             _drawingContext.DrawText(
-                text.Content,
+                text.LocalizedContent,
```
Comment (feliwir) [inline src/OpenSage.Game/Gui/Apt/AptRenderingContext.cs:138]: This is wrong

File: src/OpenSage.Game/Gui/Apt/RenderItem.cs [hunk 2]
```diff
@@ -27,6 +27,42 @@ public override void Create(Character character, AptContext context, SpriteItem
...
+                }
+                catch (System.Exception e)
+                {
+                    Logger.Warn($"Failed to resolve text value: {e}");
+                }
+
+            }
+
+            // localize our content
+            t.LocalizedContent = t.Content
```
Comment (feliwir) [inline src/OpenSage.Game/Gui/Apt/RenderItem.cs:60]: No

**Review by feliwir — CHANGES_REQUESTED** [review]

**Comment by feliwir** [comment]
@BSG-75 your changes to the Text Content do cause some issues for BFME, see the screenshot from @Tarcontar

**Comment by Tarcontar** [comment]
lookin good now in bfme and bfme2! :)

**Pull request merged**